### PR TITLE
KAFKA-7918: Inline generic parameters Pt. II: RocksDB Bytes Store and Memory LRU Caches

### PR DESCRIPTION
**Pull request by ableegoldman**
Second PR in series to inline the generic parameters of the following bytes stores:

[ Pt. I] InMemoryKeyValueStore
[x] RocksDBWindowStore
[x] RocksDBSessionStore
[x] MemoryLRUCache
[x] MemoryNavigableLRUCache
[  ] InMemoryWindowStore

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by ableegoldman** [comment]
@guozhangwang @vvcephei @mjsax @bbejeck

File: streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBWindowStoreTest.java [hunk 1]
```diff
@@ -1368,24 +1368,6 @@ public void shouldThrowNullPointerExceptionOnRangeNullToKey() {
         windowStore.fetch(1, null, ofEpochMilli(1L), ofEpochMilli(2L));
     }
 
-    @Test
-    public void shouldNoNullPointerWhenSerdeDoesNotHandleNull() {
```
Comment (ableegoldman) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBWindowStoreTest.java:1372]: Removed this test entirely as it no longer makes sense when the RocksDB store can't and shouldn't interact with the serdes at all

File: streams/src/main/java/org/apache/kafka/streams/state/internals/MemoryNavigableLRUCache.java [hunk 2]
```diff
@@ -16,44 +16,43 @@
...
+        super(name, maxCacheSize);
     }
 
     @Override
-    public KeyValueIterator<K, V> range(final K from, final K to) {
-        final TreeMap<K, V> treeMap = toTreeMap();
-        return new DelegatingPeekingKeyValueIterator<>(name(), new MemoryNavigableLRUCache.CacheIterator<>(treeMap.navigableKeySet().subSet(from, true, to, true).iterator(), treeMap));
+    public KeyValueIterator<Bytes, byte[]> range(final Bytes from, final Bytes to) {
+        final TreeMap<Bytes, byte[]> treeMap = toTreeMap();
+        return new DelegatingPeekingKeyValueIterator<>(name(), new MemoryNavigableLRUCache.CacheIterator(treeMap.navigableKeySet().subSet(from, true, to, true).iterator(), treeMap));
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/MemoryNavigableLRUCache.java]: nit: break too long line into multipl

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSessionStore.java [hunk 3]
```diff
@@ -16,90 +16,59 @@
...
     public void init(final ProcessorContext context, final StateStore root) {
-        final String storeName = name();
-        topic = ProcessorStateManager.storeChangelogTopic(context.applicationId(), storeName);
-
-        serdes = new StateSerdes<>(
-            topic,
-            keySerde == null ? (Serde<K>) context.keySerde() : keySerde,
-            aggSerde == null ? (Serde<AGG>) context.valueSerde() : aggSerde);
-
         super.init(context, root);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSessionStore.java]: If we only call `super.init()` we can remove the whole method.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSessionStore.java [hunk 4]
```diff
@@ -16,90 +16,59 @@
...
 
     @Override
-    public KeyValueIterator<Windowed<K>, AGG> findSessions(final K key, final long earliestSessionEndTime, final long latestSessionStartTime) {
-        final KeyValueIterator<Bytes, byte[]> bytesIterator = wrapped().fetch(
-            Bytes.wrap(serdes.rawKey(key)),
-            earliestSessionEndTime,
-            latestSessionStartTime
-        );
-        return new WrappedSessionStoreIterator<>(bytesIterator, serdes);
+    public KeyValueIterator<Windowed<Bytes>, byte[]> findSessions(final Bytes key, final long earliestSessionEndTime, final long latestSessionStartTime) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSessionStore.java]: break too long line into multiple (one line per parameter)

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSessionStore.java [hunk 5]
```diff
@@ -16,90 +16,59 @@
...
+        final KeyValueIterator<Bytes, byte[]> bytesIterator = wrapped().fetch(key, earliestSessionEndTime, latestSessionStartTime);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSessionStore.java]: nit: line too long (compare previous formatting)

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSessionStore.java [hunk 6]
```diff
@@ -16,90 +16,59 @@
...
     @Override
-    public KeyValueIterator<Windowed<K>, AGG> findSessions(final K keyFrom, final K keyTo, final long earliestSessionEndTime, final long latestSessionStartTime) {
-        final KeyValueIterator<Bytes, byte[]> bytesIterator = wrapped().fetch(
-            Bytes.wrap(serdes.rawKey(keyFrom)),
-            Bytes.wrap(serdes.rawKey(keyTo)),
-            earliestSessionEndTime,
-            latestSessionStartTime
-        );
-        return new WrappedSessionStoreIterator<>(bytesIterator, serdes);
+    public KeyValueIterator<Windowed<Bytes>, byte[]> findSessions(final Bytes keyFrom, final Bytes keyTo, final long earliestSessionEndTime, final long latestSessionStartTime) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSessionStore.java]: as above

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSessionStore.java [hunk 7]
```diff
@@ -16,90 +16,59 @@
...
+        final KeyValueIterator<Bytes, byte[]> bytesIterator = wrapped().fetch(keyFrom, keyTo, earliestSessionEndTime, latestSessionStartTime);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSessionStore.java]: as above

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBWindowStore.java [hunk 8]
```diff
@@ -16,98 +16,82 @@
...
+        final KeyValueIterator<Bytes, byte[]> bytesIterator = wrapped().fetch(key, timeFrom, timeTo);
+        return new WindowStoreIteratorWrapper(bytesIterator, windowSize).valuesIterator();
     }
 
     @SuppressWarnings("deprecation")
     @Override
-    public KeyValueIterator<Windowed<K>, V> fetch(final K from, final K to, final long timeFrom, final long timeTo) {
-        final KeyValueIterator<Bytes, byte[]> bytesIterator = wrapped().fetch(Bytes.wrap(serdes.rawKey(from)), Bytes.wrap(serdes.rawKey(to)), timeFrom, timeTo);
-        return new WindowStoreIteratorWrapper<>(bytesIterator, serdes, windowSize).keyValueIterator();
+    public KeyValueIterator<Windowed<Bytes>, byte[]> fetch(final Bytes from, final Bytes to, final long timeFrom, final long timeTo) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBWindowStore.java]: as above (maybe also some more below -- please double check and break every line that does not display on Github at once)

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDbWindowBytesStoreSupplier.java [hunk 9]
```diff
@@ -54,11 +53,9 @@ public String name() {
                 segmentInterval,
                 new WindowKeySchema()
         );
-        return new RocksDBWindowStore<>(segmentedBytesStore,
-                Serdes.Bytes(),
-                Serdes.ByteArray(),
-                retainDuplicates,
-                windowSize);
+        return new RocksDBWindowStore(segmentedBytesStore,
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDbWindowBytesStoreSupplier.java]: nit: move to next line and indent all with 4 spaces

**Review by mjsax — COMMENTED** [review]
LGTM. Just some formatting nits.

**Review by bbejeck — APPROVED** [review]
Thanks @ableegoldman. LGTM modulo comments from @mjsax.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBWindowStore.java (hunk 8 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBWindowStore.java]: Ack

**Comment by ableegoldman** [comment]
JDK 8 passed, JDK 11 failed on

ConsumerBounceTest.testRollingBrokerRestartsWithSmallerMaxGroupSizeConfigDisruptsBigGroup

Retest this, please

**Comment by ableegoldman** [comment]
JDK 11 failed on three unrelated tests
kafka.api.PlaintextConsumerTest.testCommitMetadata
kafka.api.PlaintextEndToEndAuthorizationTest.testNoConsumeWithDescribeAclViaAssign
kafka.server.KafkaMetricReporterClusterIdTest.testClusterIdPresent

Retest this, please

**Comment by bbejeck** [comment]
Merged #6327 to trunk

**Pull request merged**